From 929b115ec4fb5d7f3eda6ab8b3659faee3e56182 Mon Sep 17 00:00:00 2001
From: Jakub Adam <jakub.adam@ktknet.cz>
Date: Tue, 15 Mar 2011 22:26:02 +0100
Subject: [PATCH] When user rejects call, reject only sessions that were not previously
 accepted

For example when buddy adds video to running call, allow rejecting
it without dropping voice.
---
 pidgin/gtkmedia.c |    9 +++++++--
 1 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/pidgin/gtkmedia.c b/pidgin/gtkmedia.c
index ca4e10c..d45eaef 100644
--- a/pidgin/gtkmedia.c
+++ b/pidgin/gtkmedia.c
@@ -595,8 +595,13 @@ pidgin_media_accept_cb(PurpleMedia *media, int index)
 static void
 pidgin_media_reject_cb(PurpleMedia *media, int index)
 {
-	purple_media_stream_info(media, PURPLE_MEDIA_INFO_REJECT,
-			NULL, NULL, TRUE);
+	GList *iter = purple_media_get_session_ids(media);
+	for (; iter; iter = g_list_delete_link(iter, iter)) {
+		const gchar *sessionid = iter->data;
+		if (!purple_media_accepted(media, sessionid, NULL))
+			purple_media_stream_info(media, PURPLE_MEDIA_INFO_REJECT,
+					sessionid, NULL, TRUE);
+	}
 }
 
 static gboolean
-- 
1.7.2.3

